### PR TITLE
chore: Implement collector interface for EDP

### PR DESCRIPTION
**Pull request by shorim**
## Description

Changes proposed in this pull request (what was done and why):

- Implemented the collector interface for EDP
- Integrated the EDP collector to the queue processing logic
- adjusted unit tests for the queue processing and added new unit tests for EDP collector
- Removed dead code which was used before the introduction of scanners and collector

Changes refer to particular issues, PRs or documents:

- #122 

## Traceability
- [x] The PR is linked to a GitHub issue.
- [ ] The follow-up issues (if any) are linked in the `Related Issues` section.
- [ ] If the change is user-facing, the documentation has been adjusted.
- [ ] If a CRD is changed, the corresponding Busola ConfigMap has been adjusted.
- [x] The feature is unit-tested.
- [ ] The feature is e2e-tested.

### CODE REVIEW

File: pkg/logger/logger.go [hunk 1]
```diff
@@ -59,7 +59,7 @@ const (
 	// ValueTrue is used as a value for a message with true status.
 	ValueTrue = "true"
 
-	// ValueFalse is used as a value for a message with true status.
+	// ValueFalse is used as a value for a message with false status.
```
Comment (k15r) [inline pkg/logger/logger.go:62]: 😄

Comment (shorim) [reply pkg/logger/logger.go:62]: leaving the code better than you found it 😉😄

File: pkg/collector/edp/collector.go [hunk 2]
```diff
@@ -2,69 +2,204 @@ package edp
...
-			attribute.String("shoot_id", runtime.ShootInfo.ShootName),
 		),
 	)
 	defer span.End()
 
+	currentTimestamp := getTimestampNow()
+
+	scans, err := c.executeScans(childCtx, runtime)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to execute all scans: %w", err))
```
Comment (k15r) [inline pkg/collector/edp/collector.go:54]: this will print failed to execute all scans even if just a single scan failed

File: pkg/collector/edp/collector.go [hunk 3]
```diff
@@ -2,69 +2,204 @@ package edp
...
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+	}
+
+	EDPMeasurements, err := c.convertScansToEDPMeasurements(scans, previousScans, runtime)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to convert all scans to EDP measurements: %w", err))
```
Comment (k15r) [inline pkg/collector/edp/collector.go:61]: this will print failed to convert all scans even if just a single conversion failed

File: pkg/collector/edp/collector.go [hunk 4]
```diff
@@ -2,69 +2,204 @@ package edp
...
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+	}
+
+	if len(EDPMeasurements) == 0 {
+		errMsg := "no EDP measurements to send to EDP"
+		errs = append(errs, fmt.Errorf("%s", errMsg))
```
Comment (k15r) [inline pkg/collector/edp/collector.go:68]: lets extract an error for this
errNoMeasurementsSent := errors.New("no measurements sent to EDP")

File: pkg/collector/edp/collector.go [hunk 5]
```diff
@@ -2,69 +2,204 @@ package edp
...
+			if err != nil {
+				errs = append(errs, fmt.Errorf("failed to convert previous scan to an EDP measurement for scanner with ID(%s): %w", s.ID(), err))
+
+				continue
+			}
+
+			currentScans[s.ID()] = previousScan
+		} else {
+			collector.RecordScanConversion(true, string(s.ID()), collector.EDPBackendName, *runtime)
+		}
```
Comment (k15r) [inline pkg/collector/edp/collector.go:162]:
```suggestion
                success := err == nil
                collector.RecordScanConversion(success, string(s.ID()), collector.EDPBackendName, *runtime)
		if err != nil {
			errs = append(errs, fmt.Errorf("failed to convert scan to an EDP measurement for scanner with ID(%s): %w", s.ID(), err))
			previousScan, previousScanExists := previousScans[s.ID()]
			// if the previous scan doesn't exist, nothing else we can do here
			if !previousScanExists {
				errs = append(errs, fmt.Errorf("no previous scan found for scanner with ID(%s)", s.ID()))

				continue
			}

			EDPMeasurement, err = previousScan.EDP()
			// if conversion of previous scan to an EDP measurement also fails, nothing else we can do here
			if err != nil {
				errs = append(errs, fmt.Errorf("failed to convert previous scan to an EDP measurement for scanner with ID(%s): %w", s.ID(), err))

				continue
			}

			currentScans[s.ID()] = previousScan
		} 
```

File: pkg/collector/edp/collector.go [hunk 6]
```diff
@@ -2,69 +2,204 @@ package edp
...
+
+	edpRequest, err := c.EDPClient.NewRequest(subAccountID)
+	if err != nil {
+		return fmt.Errorf("failed to create a new request for EDP for subAccountID (%s): %w", subAccountID, err)
+	}
+
+	resp, err := c.EDPClient.Send(edpRequest, payloadJSON)
+	if err != nil {
+		return fmt.Errorf("failed to send payload to EDP for subAccountID (%s): %w", subAccountID, err)
+	}
```
Comment (k15r) [inline pkg/collector/edp/collector.go:185]: nit:
```suggestion
	req, err := c.EDPClient.NewRequest(subAccountID)
	if err != nil {
		return fmt.Errorf("failed to create a new request for EDP for subAccountID (%s): %w", subAccountID, err)
	}

	resp, err := c.EDPClient.Send(req, payloadJSON)
	if err != nil {
		return fmt.Errorf("failed to send payload to EDP for subAccountID (%s): %w", subAccountID, err)
	}
```

File: pkg/collector/edp/collector_test.go [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+	// scanner1 will behave the same in all test cases: successful scan and successful conversion to EDP measurement.
+	// scanner2 will behave differently in each test case:
+	// case 1: scanner2 succeeds in scanning and conversion to EDP measurement succeeds.
+	// case 2: scanner2 fails in scanning, but previous scan exists. So, the previous scan will be used for conversion to EDP measurement.
+	// case 3: scanner2 fails in scanning and previous scan doesn't exist. So, nothing else we can do here and payload will be sent to EDP without scanner2's data.
+	// case 4: scanner2 succeeds in scanning, but conversion to EDP measurement fails. So, the previous scan will be used for conversion to EDP measurement.
+	// case 5: scanner2 succeeds in scanning, but conversion to EDP measurement fails and previous scan doesn't exist. So, nothing else we can do here and payload will be sent to EDP without scanner2's data.
+	// case 6: scanner2 succeeds in scanning, but conversion to EDP measurement fails and previous scan exists, but conversion of previous scan to EDP measurement also fails. So, nothing else we can do here and payload will be sent to EDP without scanner2's data.
```
Comment (k15r) [inline pkg/collector/edp/collector_test.go:45]: is this even a valid scenario?
if conversion fails it should not be stored in the "previous" map

File: pkg/collector/edp/payload.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	Compute      resource.EDPMeasurement `json:"compute"              validate:"required"`
+	Networking   *Networking             `json:"networking,omitempty"`
+}
+
+type Networking struct {
+	ProvisionedVnets int `json:"provisioned_vnets" validate:"numeric"`
+	ProvisionedIPs   int `json:"provisioned_ips"   validate:"numeric"`
+}
+
+func newPayload(runtimeID, subAccountID, shootName, timeStamp string, EDPMeasuremnets []resource.EDPMeasurement) Payload {
```
Comment (k15r) [inline pkg/collector/edp/payload.go:19]: unexported constructor for exported type?

File: pkg/collector/edp/collector.go [hunk 9]
```diff
@@ -2,69 +2,204 @@ package edp
...
+	var errs []error
+
+	childCtx, span := otel.Tracer("").Start(ctx, "kmc.collect_scans_and_send_measurements",
 		trace.WithAttributes(
+			attribute.String("instance_id", runtime.InstanceID),
+			attribute.String("runtime_id", runtime.RuntimeID),
+			attribute.String("sub_account_id", runtime.SubAccountID),
+			attribute.String("global_account_id", runtime.GlobalAccountID),
+			attribute.String("shoot_name", runtime.ShootName),
 			attribute.String("provider", runtime.ProviderType),
```
Comment (k15r) [inline pkg/collector/edp/collector.go:45]: lets make these attribute names constants

File: pkg/collector/metrics.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+	subsystem          = "collector"
+	shootNameLabel     = "shoot_name"
+	instanceIdLabel    = "instance_id"
+	runtimeIdLabel     = "runtime_id"
+	subAccountLabel    = "sub_account_id"
+	globalAccountLabel = "global_account_id"
+	successLabel       = "success"
+	resourceNameLabel  = "resource_name"
+	backendNameLabel   = "backend_name"
+	EDPBackendName     = "edp"
```
Comment (k15r) [inline pkg/collector/metrics.go:23]: lets reuse these constants also for the traces

File: pkg/collector/edp/collector.go (hunk 2 above)
Comment (shorim) [reply pkg/collector/edp/collector.go:54]: the intention was that not all scans were executed successfully, but you are right it is misleading.
Maybe we rephrase it to  `fmt.Errorf("one or more scans failed to execute successfully: %w", err)`?

File: pkg/collector/edp/collector.go (hunk 3 above)
Comment (shorim) [reply pkg/collector/edp/collector.go:61]: also the intention here was that not all conversions were successful, but I agree it is misleading.
Maybe we rephrase it to `fmt.Errorf("failed to convert one or more scans to EDP measurements: %w", err)` ?

File: pkg/resource/pvc/scanner.go [hunk 11]
```diff
@@ -33,7 +35,16 @@ func (s *Scanner) ID() resource.ScannerID {
...
+	ctx, span := otel.Tracer("").Start(ctx, "kmc.pvc_scan",
+		trace.WithAttributes(
+			attribute.String("instance_id", runtime.InstanceID),
+			attribute.String("runtime_id", runtime.RuntimeID),
+			attribute.String("sub_account_id", runtime.SubAccountID),
+			attribute.String("global_account_id", runtime.GlobalAccountID),
+			attribute.String("shoot_name", runtime.ShootName),
+			attribute.String("provider", runtime.ProviderType),
+		),
+	)
```
Comment (k15r) [inline pkg/resource/pvc/scanner.go:47]: please extract a reusable function, as every scanner will need this

File: pkg/collector/edp/collector_test.go (hunk 7 above)
Comment (k15r) [reply pkg/collector/edp/collector_test.go:45]: lets remove this test case and verify in a test that this scenario cannot happen

File: pkg/collector/edp/payload.go (hunk 8 above)
Comment (shorim) [reply pkg/collector/edp/payload.go]: I unexported the `payload` struct as it is only used in the same package, so no need to export it.

File: pkg/collector/edp/collector.go (hunk 9 above)
Comment (shorim) [reply pkg/collector/edp/collector.go]: They are constants now in the `otel` package

File: pkg/collector/metrics.go (hunk 10 above)
Comment (shorim) [reply pkg/collector/metrics.go:23]: We have dedicated constants now in the `otel` package. So, the `SpanAttributes` helper func can be used by both collector and scanners

File: pkg/resource/pvc/scanner.go (hunk 11 above)
Comment (shorim) [reply pkg/resource/pvc/scanner.go]: We have now a reusable function `SpanAttributes` in the otel package which can be used by all scanners and collector

File: pkg/collector/edp/collector_test.go (hunk 7 above)
Comment (shorim) [reply pkg/collector/edp/collector_test.go]: I removed test case 6.
And I make sure now that the scanMap returned by `CollectAndSend` contains only scans which succeed in the 2 steps (scanning and converting).
And case 5 already ensure now that the returned scan map will not contain scan from scanner2

File: pkg/collector/edp/collector.go [hunk 12]
```diff
@@ -2,69 +2,192 @@ package edp
...
 func (c *Collector) CollectAndSend(ctx context.Context, runtime *runtime.Info, previousScans collector.ScanMap) (collector.ScanMap, error) {
-	childCtx, span := otel.Tracer("").Start(ctx, "collect",
-		trace.WithAttributes(
-			attribute.String("provider", runtime.ProviderType),
-			attribute.String("shoot_id", runtime.ShootInfo.ShootName),
-		),
-	)
+	var errs []error
+
+	childCtx, span := otel.Tracer("").Start(ctx, "kmc.collect_scans_and_send_measurements", kmcotel.SpanAttributes(runtime))
```
Comment (k15r) [inline pkg/collector/edp/collector.go:39]:
```suggestion
	childCtx, span := otel.Tracer("").Start(ctx, "collect_scans_and_send_measurements", kmcotel.SpanAttributes(runtime))
```

"kmc" should refer to the service name. The service name can be set using environment variables in the helm chart

File: pkg/collector/edp/collector.go [hunk 13]
```diff
@@ -2,69 +2,192 @@ package edp
...
+			if !previousScanExists {
+				// since even a previous scan doesn't exist, we won't be able to convert any scan to an EDP measurement, so conversion is recorded as unsuccessful
+				collector.RecordScanConversion(false, string(s.ID()), collector.EDPBackendName, *runtime)
+				errs = append(errs, fmt.Errorf("no previous scan found for scanner with ID(%s)", s.ID()))
+
+				continue
+			}
+
+			scan = previousScan
 		}
```
Comment (k15r) [inline pkg/collector/edp/collector.go:108]: how about this (remove nested ifs and makes the flow easier to understand (imho))
```
	for _, s := range c.scanners {
		scan, err := s.Scan(ctx, runtime)
		success := err == nil
		collector.RecordScan(success, string(s.ID()), *runtime)
		// if the scanner fails during execution, attempt to get the previous scan
		if success {
			currentScans[s.ID()] = scan
			continue
		}

		// we encountered an error during scanning, so we need to attempt to get the previous scan
		errs = append(errs, fmt.Errorf("scanner with ID(%s) failed during scanning: %w", s.ID(), err))

		previousScan, exists := previousScans[s.ID()]
		// if the previous scan also doesn't exist, nothing else we can do here
		if exists {
			currentScans[s.ID()] = previousScan
			continue
		}
		// since even a previous scan doesn't exist, we won't be able to convert any scan to an EDP measurement, so conversion is recorded as unsuccessful
		collector.RecordScanConversion(false, string(s.ID()), collector.EDPBackendName, *runtime)
		errs = append(errs, fmt.Errorf("no previous scan found for scanner with ID(%s)", s.ID()))

	}
```

File: pkg/collector/edp/collector.go [hunk 14]
```diff
@@ -2,69 +2,192 @@ package edp
...
+
+				continue
+			}
+
+			scan = previousScan
+		}
+
+		convertableScans[id] = scan
+		EDPMeasurements = append(EDPMeasurements, EDPMeasurement)
+	}
```
Comment (k15r) [inline pkg/collector/edp/collector.go:153]:
```suggestion
for id, scan := range currentScans {
		EDPMeasurement, err := scan.EDP()
		success := err == nil
		collector.RecordScanConversion(success, string(id), collector.EDPBackendName, *runtime)
		// if the scan can be converted to an EDP measurement, we add it to the convertableScans and EDPMeasurements
		if success {
			convertableScans[id] = scan
			EDPMeasurements = append(EDPMeasurements, EDPMeasurement)
			continue
		}

		errs = append(errs, fmt.Errorf("failed to convert scan to an EDP measurement for scanner with ID(%s): %w", string(id), err))

		// if the scan fails to convert to an EDP measurement, we attempt to get the previous scan
		previousScan, exists := previousScans[id]

		// if the previous scan doesn't exist, nothing else we can do here
		if !exists {
			errs = append(errs, fmt.Errorf("no previous scan found for scanner with ID(%s)", string(id)))

			continue
		}

		EDPMeasurement, err = previousScan.EDP()
		// if conversion of previous scan to an EDP measurement also fails, nothing else we can do here
		if err != nil {
			errs = append(errs, fmt.Errorf("failed to convert previous scan to an EDP measurement for scanner with ID(%s): %w", string(id), err))

			continue
		}

		// if the previous scan can be converted to an EDP measurement, we add it to the convertableScans and EDPMeasurements
		convertableScans[id] = previousScan
		EDPMeasurements = append(EDPMeasurements, EDPMeasurement)
	}
```

File: pkg/resource/node/scanner.go [hunk 15]
```diff
@@ -34,7 +35,7 @@ func (s *Scanner) ID() resource.ScannerID {
 }
 
 func (s *Scanner) Scan(ctx context.Context, runtime *runtime.Info) (resource.ScanConverter, error) {
-	ctx, span := otel.Tracer("").Start(ctx, "kmc.node_scan")
+	ctx, span := otel.Tracer("").Start(ctx, "kmc.node_scan", kmcotel.SpanAttributes(runtime))
```
Comment (k15r) [inline pkg/resource/node/scanner.go:38]: imho "kmc" is the service name which is set using env-variables

Comment (shorim) [reply pkg/resource/node/scanner.go]: the `kmc` prefix is now removed from the span names

**Review by k15r — APPROVED** [review]
